File: src/shipyard/core/evidence.py
"""Evidence model — per-target SHA proof tracking.

Evidence is the core of Shipyard's merge gate. It records what SHA was
validated, on what machine, by what backend, and when.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from pathlib import Path


@dataclass(frozen=True)
class EvidenceRecord:
    """Proof that a specific SHA was validated on a specific target."""

    sha: str
    branch: str
    target_name: str
    platform: str
    status: str  # "pass" or "fail"
    backend: str  # "local", "ssh", "namespace-failover", etc.
    completed_at: datetime
    duration_secs: float | None = None
    host: str | None = None
    # Failover provenance
    primary_backend: str | None = None
    failover_reason: str | None = None
    provider: str | None = None
    runner_profile: str | None = None
    # Coarse-grained failure taxonomy when ``status == "fail"``. One of
    # "INFRA" | "TIMEOUT" | "CONTRACT" | "TEST" | "UNKNOWN". None on a
    # passing record. See ``shipyard.core.classify``.
    failure_class: str | None = None

    @property
    def passed(self) -> bool:
        return self.status == "pass"

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "sha": self.sha,
            "branch": self.branch,
            "target": self.target_name,
            "platform": self.platform,
            "status": self.status,
            "backend": self.backend,
            "completed_at": self.completed_at.isoformat(),
        }
        for key in (
            "duration_secs", "host", "primary_backend", "failover_reason",
            "provider", "runner_profile", "failure_class",
        ):
            val = getattr(self, key)
            if val is not None:
                d[key] = val
        return d

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> EvidenceRecord:
        return cls(
            sha=d["sha"],
            branch=d["branch"],
            target_name=d["target"],
            platform=d["platform"],
            status=d["status"],
            backend=d["backend"],
            completed_at=datetime.fromisoformat(d["completed_at"]),
            duration_secs=d.get("duration_secs"),
            host=d.get("host"),
            primary_backend=d.get("primary_backend"),
            failover_reason=d.get("failover_reason"),
            provider=d.get("provider"),
            runner_profile=d.get("runner_profile"),
            failure_class=d.get("failure_class"),
        )


@dataclass
class EvidenceStore:
    """Persists and queries evidence records.

    Evidence is stored per-branch, keyed by target name. Only the most
    recent record per target is kept (you only need to know the latest
    proof for merge gating).
    """

    path: Path
    _cache: dict[str, dict[str, EvidenceRecord]] = field(default_factory=dict, repr=False)

    def __post_init__(self) -> None:
        self.path.mkdir(parents=True, exist_ok=True)

    def record(self, evidence: EvidenceRecord) -> None:
        """Store an evidence record, replacing any existing for same branch+target."""
        branch_key = _sanitize_branch(evidence.branch)
        branch_data = self._load_branch(branch_key)
        branch_data[evidence.target_name] = evidence
        self._save_branch(branch_key, branch_data)

    def get_branch(self, branch: str) -> dict[str, EvidenceRecord]:
        """Get all evidence for a branch, keyed by target name."""
        return dict(self._load_branch(_sanitize_branch(branch)))

    def get_target(self, branch: str, target_name: str) -> EvidenceRecord | None:
        """Get evidence for a specific branch + target."""
        return self._load_branch(_sanitize_branch(branch)).get(target_name)

    def is_merge_ready(
        self,
        branch: str,
        sha: str,
        required_platforms: list[str],
    ) -> tuple[bool, dict[str, EvidenceRecord | None]]:
        """Check if all required platforms have passing evidence for this SHA.

        Returns (ready, evidence_map) where evidence_map has an entry for
        each required platform (None if no evidence exists).
        """
        records = self.get_branch(branch)
        evidence_map: dict[str, EvidenceRecord | None] = {}
        all_green = True

        for platform in required_platforms:
            # Find evidence for this platform
            match = None
            for rec in records.values():
                if rec.platform == platform and rec.sha == sha and rec.passed:
                    match = rec
                    break
            evidence_map[platform] = match
            if match is None:
                all_green = False

        return all_green, evidence_map

    def _branch_file(self, branch_key: str) -> Path:
        return self.path / f"{branch_key}.json"

    def _load_branch(self, branch_key: str) -> dict[str, EvidenceRecord]:
        if branch_key in self._cache:
            return self._cache[branch_key]

        path = self._branch_file(branch_key)
        if not path.exists():
            self._cache[branch_key] = {}
            return self._cache[branch_key]

        data = json.loads(path.read_text())
        records = {k: EvidenceRecord.from_dict(v) for k, v in data.items()}
        self._cache[branch_key] = records
        return records

    def _save_branch(self, branch_key: str, records: dict[str, EvidenceRecord]) -> None:
        self._cache[branch_key] = records
        path = self._branch_file(branch_key)
        data = {k: v.to_dict() for k, v in records.items()}
        path.write_text(json.dumps(data, indent=2) + "\n")


def _sanitize_branch(branch: str) -> str:
    """Convert branch name to a safe filename."""
    return branch.replace("/", "--").replace("\\", "--")

File: src/evidence.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Proof that a specific SHA was validated on a specific target.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct EvidenceRecord {
    /// Validated git SHA.
    pub sha: String,
    /// Git branch associated with the run.
    pub branch: String,
    /// Logical target name.
    #[serde(rename = "target")]
    pub target_name: String,
    /// Concrete platform label.
    pub platform: String,
    /// Validation result status.
    pub status: String,
    /// Backend that produced this evidence.
    pub backend: String,
    /// Completion timestamp.
    pub completed_at: DateTime<Utc>,
    /// Optional duration in seconds.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_secs: Option<f64>,
    /// Optional host identifier.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub host: Option<String>,
    /// Primary backend when failover occurred.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub primary_backend: Option<String>,
    /// Failover reason.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub failover_reason: Option<String>,
    /// Cloud provider label.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    /// Runner profile label.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub runner_profile: Option<String>,
    /// Coarse failure class.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub failure_class: Option<String>,
    /// Ancestor SHA this record reused.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reused_from: Option<String>,
    /// Digest of the contract in effect when the record was written.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub contract_digest: Option<String>,
    /// Stable signature of the stage pipeline.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub stages_signature: Option<String>,
}

impl EvidenceRecord {
    /// Whether this record is a passing validation.
    #[must_use]
    pub fn passed(&self) -> bool {
        self.status == "pass"
    }

    /// Whether this record was synthesized from an ancestor pass.
    #[must_use]
    pub fn reused(&self) -> bool {
        self.reused_from.is_some()
    }
}

/// Persistent store for per-branch evidence records.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct EvidenceStore {
    path: PathBuf,
}

impl EvidenceStore {
    /// Open an evidence store at the given path.
    pub fn new(path: PathBuf) -> Result<Self, std::io::Error> {
        fs::create_dir_all(&path)?;
        Ok(Self { path })
    }

    /// Backing path of the store.
    #[must_use]
    pub fn path(&self) -> &Path {
        &self.path
    }

    /// Store or replace a record for the same branch and target.
    pub fn record(&self, evidence: &EvidenceRecord) -> Result<(), Box<dyn std::error::Error>> {
        let branch_key = sanitize_branch(&evidence.branch);
        let mut records = self.load_branch(&branch_key)?;
        records.insert(evidence.target_name.clone(), evidence.clone());
        self.save_branch(&branch_key, &records)
    }

    /// Return all evidence for a branch keyed by target name.
    #[must_use]
    pub fn get_branch(&self, branch: &str) -> BTreeMap<String, EvidenceRecord> {
        self.load_branch(&sanitize_branch(branch))
            .unwrap_or_default()
    }

    /// Return evidence for a specific branch and target, if present.
    #[must_use]
    pub fn get_target(&self, branch: &str, target_name: &str) -> Option<EvidenceRecord> {
        self.get_branch(branch).remove(target_name)
    }

    /// Return whether every required platform has passing evidence for the SHA.
    #[must_use]
    pub fn is_merge_ready(
        &self,
        branch: &str,
        sha: &str,
        required_platforms: &[String],
    ) -> (bool, BTreeMap<String, Option<EvidenceRecord>>) {
        let records = self.get_branch(branch);
        let mut evidence_map = BTreeMap::new();
        let mut all_green = true;

        for platform in required_platforms {
            let record = records
                .values()
                .find(|record| record.platform == *platform && record.sha == sha && record.passed())
                .cloned();
            if record.is_none() {
                all_green = false;
            }
            evidence_map.insert(platform.clone(), record);
        }

        (all_green, evidence_map)
    }

    /// Find the highest-ranked passing record for a target across all branches.
    #[must_use]
    pub fn query_passing_for_target(
        &self,
        target_name: &str,
        sha_candidates: &[String],
    ) -> Option<EvidenceRecord> {
        let candidate_ranks = sha_candidates
            .iter()
            .enumerate()
            .map(|(rank, sha)| (sha.as_str(), rank))
            .collect::<BTreeMap<_, _>>();
        let mut best: Option<(usize, EvidenceRecord)> = None;

        let Ok(entries) = fs::read_dir(&self.path) else {
            return None;
        };

        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().and_then(std::ffi::OsStr::to_str) != Some("json") {
                continue;
            }
            let Some(branch_key) = path.file_stem().and_then(std::ffi::OsStr::to_str) else {
                continue;
            };
            let Ok(records) = self.load_branch(branch_key) else {
                continue;
            };
            for record in records.values() {
                if record.target_name != target_name || !record.passed() || record.reused() {
                    continue;
                }
                let Some(rank) = candidate_ranks.get(record.sha.as_str()).copied() else {
                    continue;
                };
                if best.as_ref().is_none_or(|(best_rank, _)| rank < *best_rank) {
                    best = Some((rank, record.clone()));
                }
            }
        }

        best.map(|(_, record)| record)
    }

    fn branch_file(&self, branch_key: &str) -> PathBuf {
        self.path.join(format!("{branch_key}.json"))
    }

    fn load_branch(
        &self,
        branch_key: &str,
    ) -> Result<BTreeMap<String, EvidenceRecord>, Box<dyn std::error::Error>> {
        let path = self.branch_file(branch_key);
        if !path.exists() {
            return Ok(BTreeMap::new());
        }
        let contents = fs::read_to_string(path)?;
        Ok(serde_json::from_str(&contents)?)
    }

    fn save_branch(
        &self,
        branch_key: &str,
        records: &BTreeMap<String, EvidenceRecord>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let payload = serde_json::to_string_pretty(records)?;
        let temp = tempfile::NamedTempFile::new_in(&self.path)?;
        fs::write(temp.path(), format!("{payload}\n"))?;
        temp.persist(self.branch_file(branch_key))?;
        Ok(())
    }
}

fn sanitize_branch(branch: &str) -> String {
    branch.replace(['/', '\\'], "--")
}

#[cfg(test)]
mod tests {
    use chrono::Utc;

    use super::{EvidenceRecord, EvidenceStore};

    fn record(branch: &str, target: &str, sha: &str) -> EvidenceRecord {
        EvidenceRecord {
            sha: sha.to_owned(),
            branch: branch.to_owned(),
            target_name: target.to_owned(),
            platform: format!("{target}-platform"),
            status: "pass".to_owned(),
            backend: "local".to_owned(),
            completed_at: Utc::now(),
            duration_secs: None,
            host: None,
            primary_backend: None,
            failover_reason: None,
            provider: None,
            runner_profile: None,
            failure_class: None,
            reused_from: None,
            contract_digest: None,
            stages_signature: None,
        }
    }

    #[test]
    fn evidence_record_round_trips_reuse_fields() {
        let record = EvidenceRecord {
            sha: "new".to_owned(),
            branch: "feat/x".to_owned(),
            target_name: "mac".to_owned(),
            platform: "macos-arm64".to_owned(),
            status: "pass".to_owned(),
            backend: "reused".to_owned(),
            completed_at: Utc::now(),
            duration_secs: None,
            host: None,
            primary_backend: None,
            failover_reason: None,
            provider: None,
            runner_profile: None,
            failure_class: None,
            reused_from: Some("old".to_owned()),
            contract_digest: Some("abc123".to_owned()),
            stages_signature: Some("build|test".to_owned()),
        };

        assert!(record.passed());
        assert!(record.reused());

        let value = serde_json::to_value(&record).expect("serialize");
        assert_eq!(value["target"], "mac");
        assert_eq!(value["reused_from"], "old");
        assert_eq!(value["contract_digest"], "abc123");
        assert_eq!(value["stages_signature"], "build|test");

        let restored: EvidenceRecord = serde_json::from_value(value).expect("deserialize");
        assert_eq!(restored, record);
    }

    #[test]
    fn evidence_record_omits_reuse_fields_when_absent() {
        let record = record("main", "mac", "abc");
        let value = serde_json::to_value(&record).expect("serialize");
        assert!(value.get("reused_from").is_none());
        assert!(value.get("target_name").is_none());
    }

    #[test]
    fn record_and_retrieve_branch_data() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = EvidenceStore::new(temp.path().join("evidence")).expect("store");
        let record = record("feat/x", "mac", "abc");

        store.record(&record).expect("record");

        assert_eq!(
            store.get_target("feat/x", "mac").expect("record").sha,
            "abc"
        );
        assert_eq!(store.get_branch("feat/x").len(), 1);
    }

    #[test]
    fn latest_record_overwrites_by_target() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = EvidenceStore::new(temp.path().join("evidence")).expect("store");

        store.record(&record("main", "mac", "old")).expect("record");
        store.record(&record("main", "mac", "new")).expect("record");

        assert_eq!(store.get_target("main", "mac").expect("record").sha, "new");
    }

    #[test]
    fn branch_names_are_safely_sanitized() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = EvidenceStore::new(temp.path().join("evidence")).expect("store");
        store
            .record(&record("feat/x\\nested", "mac", "abc"))
            .expect("record");

        assert!(store.path().join("feat--x--nested.json").exists());
        assert!(store.get_target("feat/x\\nested", "mac").is_some());
    }

    #[test]
    fn merge_ready_requires_all_required_platforms() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = EvidenceStore::new(temp.path().join("evidence")).expect("store");
        let mut mac = record("main", "mac", "abc");
        mac.platform = "macos-arm64".to_owned();
        let mut linux = record("main", "linux", "abc");
        linux.platform = "linux-x64".to_owned();
        store.record(&mac).expect("record");
        store.record(&linux).expect("record");

        let (ready, evidence) = store.is_merge_ready(
            "main",
            "abc",
            &["macos-arm64".to_owned(), "linux-x64".to_owned()],
        );

        assert!(ready);
        assert!(evidence.values().all(Option::is_some));
    }

    #[test]
    fn merge_ready_rejects_missing_wrong_or_failed_records() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = EvidenceStore::new(temp.path().join("evidence")).expect("store");

        let mut wrong_sha = record("main", "mac", "old");
        wrong_sha.platform = "macos-arm64".to_owned();
        let mut failed = record("main", "linux", "abc");
        failed.platform = "linux-x64".to_owned();
        failed.status = "fail".to_owned();
        store.record(&wrong_sha).expect("record");
        store.record(&failed).expect("record");

        let (ready, evidence) = store.is_merge_ready(
            "main",
            "abc",
            &["macos-arm64".to_owned(), "linux-x64".to_owned()],
        );

        assert!(!ready);
        assert!(evidence["macos-arm64"].is_none());
        assert!(evidence["linux-x64"].is_none());
    }

    #[test]
    fn store_persists_across_instances() {
        let temp = tempfile::tempdir().expect("tempdir");
        let path = temp.path().join("evidence");
        let store = EvidenceStore::new(path.clone()).expect("store");
        store.record(&record("main", "mac", "abc")).expect("record");

        let reopened = EvidenceStore::new(path).expect("store");
        assert_eq!(
            reopened.get_target("main", "mac").expect("record").sha,
            "abc"
        );
    }

    #[test]
    fn query_passing_for_target_uses_candidate_rank_and_filters_invalid_records() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = EvidenceStore::new(temp.path().join("evidence")).expect("store");

        store
            .record(&record("ba", "mac", &"a".repeat(40)))
            .expect("record");
        store
            .record(&record("bb", "mac", &"b".repeat(40)))
            .expect("record");
        store
            .record(&record("bc", "mac", &"c".repeat(40)))
            .expect("record");

        let mut reused = record("main", "mac", "abc");
        reused.backend = "reused".to_owned();
        reused.reused_from = Some("parent".to_owned());
        store.record(&reused).expect("record");

        let mut failed = record("main", "linux", "abc");
        failed.status = "fail".to_owned();
        store.record(&failed).expect("record");

        let candidates = vec!["b".repeat(40), "c".repeat(40), "a".repeat(40)];
        let match_record = store
            .query_passing_for_target("mac", &candidates)
            .expect("record");
        assert_eq!(match_record.sha, "b".repeat(40));
        assert!(
            store
                .query_passing_for_target("linux", &["abc".to_owned()])
                .is_none()
        );
        assert!(
            store
                .query_passing_for_target("unknown", &["abc".to_owned()])
                .is_none()
        );
    }
}
